### ISSUE TITLE
Crash when hg branch name is /?

### ISSUE THREAD

**Issue by ermshiperete** [frej/fast-export#91]
When hg-fast-export encounters the branch name '/?' it crashes.

hg-fast-export deals with some branch names that are invalid in git:

    Hg branch  -> git branch 
    a? -> a_
    a/ -> a_
    a/b -> a/b
    a/? -> a/_
    ?a -> _a
    a. -> a_
    a.b -> a_b
    .a -> _a
    / -> _

hg-fast-export crashes when the mercurial repo contains branch names starting with a `/` (unless the name is '/'):

    /a

Even providing a branch mapping file doesn't help.

**Comment by ermshiperete** [comment]
Stacktrace with latest master branch:

	Loaded 146 authors
	Traceback (most recent call last):
	  File "$HOME/bin/fast-export/hg-fast-export.py", line 504, in <module>
	    notes=options.notes,encoding=encoding,fn_encoding=fn_encoding))
	  File "$HOME/bin/fast-export/hg-fast-export.py", line 382, in hg2git
	    if not verify_heads(ui,repo,heads_cache,force,branchesmap):
	  File "$HOME/bin/fast-export/hg-fast-export.py", line 340, in verify_heads
	    sanitized_name=sanitize_name(b,"branch",branchesmap)
	  File "$HOME/bin/fast-export/hg-fast-export.py", line 168, in sanitize_name
	    n='/'.join(map(dot,n.split('/')))
	  File "$HOME/bin/fast-export/hg-fast-export.py", line 161, in dot
	    if name[0] == '.': return '_'+name[1:]
	IndexError: string index out of range

**Comment by frej** [comment]
I created a dummy hg repo containing branches named as in your initial report. Running with a branch mapping file (-B) containing:

```text
a?=valid-0
a/=valid-1
a/b=valid-2
a/?=valid-3
?a=valid-4
a.=valid-5
a.b=valid-6
.a=valid-7
/=valid-8
```
works fine for me. Please provide a step-by step instruction for creating your troublesome hg repo, the contents of your mapping file, and the exact command line you use to run fast-export.

**Comment by ermshiperete** [comment]
As you wish:

	~/$ cd ~/bin/fast-export/
	~/bin/fast-export[remotes/origin/HEAD]/$ git log -1
	commit fb05ce5b7bb549ef5dbfeaa4ec5740b78c4adcbe
	Author: Emilian Bold <emilian.bold@gmail.com>
	Date:   Fri Dec 30 21:47:01 2016 +0200

	    Show warning when core.ignoreCase is true
	    
	    When core.ignoreCase is true, which it is by default on OSX,
	    fast-import will produce empty changesets for renames that just change
	    the case of the file name. As this most probably is not the desired
	    behavior, trigger an error but allow the conversion to proceed if
	    --force is used.
	~/bin/fast-export[remotes/origin/HEAD]/$ python --version
	Python 2.7.12
	~/bin/fast-export[remotes/origin/HEAD]/$ lsb_release -a
	No LSB modules are available.
	Distributor ID:	Ubuntu
	Description:	Ubuntu 16.04.2 LTS
	Release:	16.04
	Codename:	xenial
	~/bin/fast-export[remotes/origin/HEAD]/$ cd /tmp
	/tmp/$ hg init issue-91
	/tmp/$ cd issue-91/
	bash: /tmp/issue-91/.hg/branch: No such file or directory
	/tmp/issue-91[]/$ hg branch /?
	marked working directory as branch /?
	(branches are permanent and global, did you want a bookmark?)
	/tmp/issue-91[/?]/$ touch foo && hg commit -m "foo"
	/tmp/issue-91[/?]/$ cat /tmp/issue91.map 
	a?=valid-0
	a/=valid-1
	a/b=valid-2
	a/?=valid-3
	?a=valid-4
	a.=valid-5
	a.b=valid-6
	.a=valid-7
	/=valid-8
	/tmp/issue-91[/?]/$ cd ..
	/tmp/$ git init issue-91.git
	Initialized empty Git repository in /tmp/issue-91.git/.git/
	/tmp/$ cd issue-91.git/
	/tmp/issue-91.git[master]/$ ~/bin/fast-export/hg-fast-export.sh -r /tmp/issue-91/ -B /tmp/issue91.map
	Loaded 9 branches
	Traceback (most recent call last):
	  File "~/bin/fast-export/hg-fast-export.py", line 504, in <module>
	    notes=options.notes,encoding=encoding,fn_encoding=fn_encoding))
	  File "~/bin/fast-export/hg-fast-export.py", line 382, in hg2git
	    if not verify_heads(ui,repo,heads_cache,force,branchesmap):
	  File "~/bin/fast-export/hg-fast-export.py", line 340, in verify_heads
	    sanitized_name=sanitize_name(b,"branch",branchesmap)
	  File "~/bin/fast-export/hg-fast-export.py", line 168, in sanitize_name
	    n='/'.join(map(dot,n.split('/')))
	  File "~/bin/fast-export/hg-fast-export.py", line 161, in dot
	    if name[0] == '.': return '_'+name[1:]
	IndexError: string index out of range
	git-fast-import statistics:
	---------------------------------------------------------------------
	Alloc'd objects:       5000
	Total objects:            0 (         0 duplicates                  )
	      blobs  :            0 (         0 duplicates          0 deltas of          0 attempts)
	      trees  :            0 (         0 duplicates          0 deltas of          0 attempts)
	      commits:            0 (         0 duplicates          0 deltas of          0 attempts)
	      tags   :            0 (         0 duplicates          0 deltas of          0 attempts)
	Total branches:           0 (         0 loads     )
	      marks:           1024 (         0 unique    )
	      atoms:              0
	Memory total:          2282 KiB
	       pools:          2048 KiB
	     objects:           234 KiB
	---------------------------------------------------------------------
	pack_report: getpagesize()            =       4096
	pack_report: core.packedGitWindowSize = 1073741824
	pack_report: core.packedGitLimit      = 8589934592
	pack_report: pack_used_ctr            =          0
	pack_report: pack_mmap_calls          =          0
	pack_report: pack_open_windows        =          0 /          0
	pack_report: pack_mapped              =          0 /          0
	---------------------------------------------------------------------

	/tmp/issue-91.git[master]/$

**Comment by frej** [comment]
There are two problems here, first you have no mapping for `/?` so that's why providing a mapping file doesn't help. Without the mapping fast-export will hit the default transform and `/?` will be transformed to `/_`, then we'll split `/_` into `['', '_']` by `n.split('/')`, and then map it with
`dot()`. In `dot()` we hit the second problem, `dot()` does not handle
the empty string.

Add a mapping for `/?` and it will start working, I'll cook up a patch
which makes `dot()` handle the empty string. That will stop us from
crashing in fast-export when a branch starts with `/`.

**Comment by ermshiperete** [comment]
Yes, you're right. After adding a mapping for `/?` things work. Don't know what I did wrong when I did wrong when I tried that the first time. Thanks!

**Comment by ermshiperete** [comment]
Crash fixed in 4bb50bb3fba1b86decf62d1f8763d56e34ac902c; importing `/?` (or any other branch starting with `/`) will require a mapping file.
